Add Totalconnect config flow

austinmroczek opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds a config flow for totalconnect so users can enter config via the user interface.  Adds tests for it as well.

Bumps total-connect-client from 0.50 to 0.52.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
# YAML configuration is no longer required..but if you want to
totalconnect:
  username: !secret totalconnect_username
  password: !secret totalconnect_password

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
https://github.com/home-assistant/home-assistant.io/pull/12171

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [n/a] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->
Working toward this...maybe next PR
- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

austinmroczek committed:
Merge pull request #1 from home-assistant/dev

update to latest HA

austinmroczek committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into dev

austinmroczek committed:
Bump skybellpy to 0.4.0

austinmroczek committed:
Bump skybellpy to 0.4.0 in requirements_all.txt

austinmroczek committed:
Added extra states for STATE_ALARM_TRIGGERED to allow users to know if
it is a burglar or fire or carbon monoxide so automations can take
appropriate actions.  Updated TotalConnect component to handle these new
states.

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant.git into dev

austinmroczek committed:
Merge pull request #2 from home-assistant/dev

update

austinmroczek committed:
Fix const import

austinmroczek committed:
Fix const import

austinmroczek committed:
Fix const imports

austinmroczek committed:
Bump total-connect-client to 0.26.

austinmroczek committed:
Catch details of alarm trigger in state attributes.

Also bumps total_connect_client to 0.27.

austinmroczek committed:
Merge branch 'dev' of
https://github.com/austinmroczek/home-assistant.git into dev

austinmroczek committed:
Change state_attributes() to device_state_attributes()

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant into dev

austinmroczek committed:
Move totalconnect component toward being a multi-platform integration.  Bump total_connect_client to 0.28.

austinmroczek committed:
Merge pull request #4 from home-assistant/dev

pull in latest

austinmroczek committed:
add missing total-connect alarm state mappings

austinmroczek committed:
Made recommended changes of MartinHjelmare at
https://github.com/home-assistant/home-assistant/pull/24427

austinmroczek committed:
Update __init__.py

austinmroczek committed:
Updates per MartinHjelmare comments

austinmroczek committed:
flake8/pydocstyle fixes

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant.git into dev

austinmroczek committed:
removed . at end of log message

austinmroczek committed:
added blank line between logging and voluptuous

austinmroczek committed:
more fixes

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Adding totalconnect zones as HA binary_sensors

austinmroczek committed:
fix manifest.json

austinmroczek committed:
flake8/pydocstyle fixes.  Added codeowner.

austinmroczek committed:
Update formatting per @springstan guidance.

austinmroczek committed:
Fixed pylint

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Add zone ID to log message for easier troubleshooting

austinmroczek committed:
Account for bypassed zones in update()

austinmroczek committed:
More status handling fixes.

austinmroczek committed:
Fixed flake8 error

austinmroczek committed:
Another attempt at black/isort fixes.

austinmroczek committed:
Bump total-connect-client to 0.50.  Simplify code using new functions in
total-connect-client package instead of importing constants.  Run black
and isort.

austinmroczek committed:
Fix manifest file

austinmroczek committed:
Another manifest fix

austinmroczek committed:
one more manifest fix

austinmroczek committed:
more manifest changes.

austinmroczek committed:
sync up

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Merge branch 'dev' of https://github.com/austinmroczek/home-assistant into dev

austinmroczek committed:
fix indent

austinmroczek committed:
one more pylint fix

austinmroczek committed:
Hopefully the last pylint fix

austinmroczek committed:
make variable names understandable

austinmroczek committed:
create and fill dict in one step

austinmroczek committed:
Fix name and attributes

austinmroczek committed:
rename to logical variable in alarm_control_panel

austinmroczek committed:
Remove location_name from alarm_control_panel attributes since it is
already the name of the alarm.

austinmroczek committed:
Multiple fixes to improve code per @springstan suggestions

austinmroczek committed:
Update homeassistant/components/totalconnect/binary_sensor.py

Co-Authored-By: springstan <46536646+springstan@users.noreply.github.com>

austinmroczek committed:
Multiple changes per @MartinHjelmare review

austinmroczek committed:
simplify alarm adding

austinmroczek committed:
Fix binary_sensor.py is_on

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
Move DOMAIN to .const in line with examples.

austinmroczek committed:
Move to async_setup

austinmroczek committed:
Simplify code using new features of total-connect-client 0.51

austinmroczek committed:
First crack at config flow for totalconnect

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into dev

austinmroczek committed:
bump totalconnect to 0.52

austinmroczek committed:
use client.is_logged_in() to avoid total-connect-client details.

austinmroczek committed:
updated generated/config_flow.py

austinmroczek committed:
use is_logged_in()

austinmroczek committed:
Hopefully final touches for config flow

austinmroczek committed:
Add tests for config flow

austinmroczek committed:
Updated requirements for test

austinmroczek committed:
Fixes to test_config_flow

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into totalconnect-config-flow

austinmroczek committed:
Removed leftover comments and code

austinmroczek committed:
fix const.py flake8 error

Kane610 commented on homeassistant/components/totalconnect/.translations/en.json:
```diff
@@ -0,0 +1,20 @@
+                    "password": "Password",
+                    "username": "Username"
+                },
+                "title": "Fill in your Total Connect login information"
```
There is also a description field that might better suite this information

austinmroczek replied:
Simplified this text.

Kane610 commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -24,30 +27,64 @@
-    password = conf[CONF_PASSWORD]
+    conf = entry.data
+    username = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASSWORD)
```
Aren't these mandatory?

austinmroczek replied:
Removed .get()

Kane610 commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -24,30 +27,64 @@
+    username = conf.get(CONF_USERNAME)
+    password = conf.get(CONF_PASSWORD)
 
     client = TotalConnectClient.TotalConnectClient(username, password)
```
This is doing I/O, which is not allowed in async context

austinmroczek replied:
Think this is fixed now.

Kane610 commented on homeassistant/components/totalconnect/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+        if user_input is not None:
+            # Validate user input
+            username = user_input.get(CONF_USERNAME)
+            password = user_input.get(CONF_PASSWORD)
```
These are required so should always be available

austinmroczek replied:
removed .get()

Kane610 commented on homeassistant/components/totalconnect/config_flow.py:
```diff
@@ -0,0 +1,59 @@
+
+    async def is_valid(self, username="", password=""):
+        """Return true if the given username and password are valid."""
+        client = TotalConnectClient.TotalConnectClient(username, password)
```
I/O in async context not allowed

austinmroczek replied:
@Kane610 please help.  I don't understand what needs to be done here.  I need to do I/O to ensure the username and password are valid.  This follows the Hue component which is given as an example.  Do I need to put this I/O inside a hass.async_add_executor_job() call?

Kane610 replied:
Exactly, check out axis component which also has a sync library

austinmroczek replied:
Thanks.  Think this is fixed now.

Kane610 commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,79 @@
+        context={"source": SOURCE_USER},
+        data={CONF_USERNAME: USERNAME, CONF_PASSWORD: PASSWORD},
+    )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
```
This test looks to allow the TotalConnect client to do I/O, that should be mocked

austinmroczek replied:
think this is complete.

Kane610 reviewed (changes requested):
You can't use the totalconnect library as is since it is sync using requests. Some changes needs to be made here

austinmroczek committed:
Simplify text per @Kane610 https://github.com/home-assistant/home-assistant/pull/32126#pullrequestreview-364652949

austinmroczek committed:
Remove .get() to speed things up since the required items should always
be available.

austinmroczek committed:
Merge branch 'totalconnect-config-flow' of https://github.com/austinmroczek/home-assistant into totalconnect-config-flow

austinmroczek committed:
Move CONF_USERNAME and CONF_PASSWORD into .const to eliminate extra I/O
to import from homeassistant.const

austinmroczek committed:
Merge branch 'totalconnect-config-flow' of https://github.com/austinmroczek/home-assistant.git into totalconnect-config-flow

austinmroczek committed:
Fix I/O async issues

austinmroczek committed:
Fix flake8 and black errors

austinmroczek committed:
Mock the I/O in tests.

austinmroczek committed:
Fix isort error

austinmroczek commented:
@Kane610 thanks for the pointers. I think this is ready for another look.  

I'm getting an Azure pipeline error related to `codecov` that I don't think is my fault.

austinmroczek committed:
Empty commit to re-start azure pipelines (per discord)

austinmroczek committed:
bump total-connect-client to 0.53

austinmroczek commented:
OK figured out the pipeline issue.  Ready for a final review.

austinmroczek reviewed (commented):
reviewed all

balloob commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -24,30 +26,56 @@
 
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Old way to set up."""
+    return True
```
I've added the breaking change label.

In the description you've added that configuration.yaml is no longer required, but this code does not support importing it.

austinmroczek replied:
I'll go with breaking change.  Updated the PR info above and the linked documentation PR.

balloob commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -24,30 +26,56 @@
         return False
 
-    hass.data[DOMAIN] = TotalConnectSystem(username, password, client)
+    hass.data[DOMAIN][username] = TotalConnectSystem(username, password, client)
```
Instead of username, let's use `entry.entry_id`
```suggestion
    hass.data[DOMAIN][entry.entry_id] = TotalConnectSystem(username, password, client)
```

balloob commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -24,30 +26,56 @@
+    )
 
-    if client.token is False:
+    if not client.is_logged_in():
```
This is doing I/O and so also has to be run in the executor.

balloob commented on homeassistant/components/totalconnect/const.py:
```diff
@@ -0,0 +1,5 @@
+"""TotalConnect constants."""
+
+DOMAIN = "totalconnect"
+CONF_PASSWORD = "password"
```
No need to define them yourself. Instead import them from `homeassistant.const`. Same for username.

austinmroczek replied:
done

balloob replied:
Please remove them from here then.

balloob commented on homeassistant/components/totalconnect/config_flow.py:
```diff
@@ -0,0 +1,60 @@
+            if valid:
+                # authentication success / valid
+                return self.async_create_entry(
+                    title="Total Connect",
```
We always put the integration name in front of the title, so this should identify the account.
```suggestion
                    title=username,
```

balloob reviewed (approved):
This looks great. It's almost done, a few small comments.

austinmroczek committed:
Update homeassistant/components/totalconnect/__init__.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

austinmroczek committed:
Update homeassistant/components/totalconnect/config_flow.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

austinmroczek committed:
Fixes per @balloob comments

austinmroczek committed:
Fix imports

austinmroczek committed:
fix isort error

austinmroczek committed:
Fix async_unload_entry

It still referenced CONF_USERNAME instead of entry.entity_id

balloob commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -1,16 +1,19 @@
 
-DOMAIN = "totalconnect"
 
 CONFIG_SCHEMA = vol.Schema(
```
You can remove this now.

austinmroczek replied:
removed

balloob commented on homeassistant/components/totalconnect/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+            step_id="user", data_schema=data_schema, errors=errors
+        )
+
+    async def async_step_import(self, user_input):
```
This import step is not used, you can remove it. You can also use it by adding this to `async_setup` in `__init__`:

```yaml
if DOMAIN not in config:
    return True

hass.async_create_task(
    hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": config_entries.SOURCE_IMPORT},
        data=config[DOMAIN],
    )
)

return True
```

That way it won't be a breaking change.

austinmroczek replied:
Took your suggestion to keep the import.  Thanks

austinmroczek committed:
Added async_setup so not breaking change.  Fixed imports.

MartinHjelmare commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+    with patch(
+        "homeassistant.components.totalconnect.config_flow.TotalConnectConfigFlow.is_valid",
+        return_value=mock_coro(True),
+    ):
```
Please patch the library code instead of our code in the config flow. We want to test the config flow code.
```suggestion
        "homeassistant.components.totalconnect.config_flow.TotalConnectClient.TotalConnectClient"
    ) as client_mock:
    client_mock.return_value.is_logged_in.return_value = True
```

austinmroczek replied:
good point, thanks

MartinHjelmare commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -23,31 +29,66 @@
+    )
 
-    if client.token is False:
+    if not client.is_logged_in():
```
Does this do blocking I/O?

austinmroczek replied:
good point, thanks.

austinmroczek replied:
It's just a flag in memory

MartinHjelmare commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -23,31 +29,66 @@
         return False
 
-    hass.data[DOMAIN] = TotalConnectSystem(username, password, client)
+    hass.data[DOMAIN][entry.entry_id] = TotalConnectSystem(username, password, client)
```
We don't seem to use username and password on this instance.

MartinHjelmare replied:
Maybe we can drop this instance all together?

austinmroczek replied:
Good point.  Sorry I blindly followed an example and didn't stop to think about it.

austinmroczek committed:
Update tests/components/totalconnect/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

austinmroczek committed:
Remove TotalConnectSystem() per @MartinHjelmare suggestion

austinmroczek committed:
Moved from is_logged_in() to is_valid_credentials()

The second is more accurate for what we are checking for, because is_logged_in() could return False due to connection error.

austinmroczek committed:
Fix import in test

austinmroczek commented:
Think this is ready now, but getting a weird azure pipeline error that is not related to my code.

MartinHjelmare commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -1,17 +1,23 @@
+
+# import voluptuous as vol
+
+# import homeassistant.helpers.config_validation as cv
```
Commented code.

austinmroczek replied:
removed

MartinHjelmare commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -1,17 +1,23 @@
 
+_LOGGER = logging.getLogger(__name__)
+
+"""
```
I don't think the config schema should be a string.

austinmroczek replied:
removed

MartinHjelmare commented on homeassistant/components/totalconnect/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
```
Remove the decorator.

austinmroczek replied:
removed

MartinHjelmare commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,105 @@
+    """Test user config."""
+    # no data provided so show the form
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=None
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

MartinHjelmare commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,105 @@
+async def test_import(hass):
+    """Test import step with good username and password."""
+    with patch(
+        "homeassistant.components.totalconnect.config_flow.TotalConnectConfigFlow.is_valid",
```
Change this too.

austinmroczek replied:
done

MartinHjelmare commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,105 @@
+
+    # Should fail, same USERNAME (import)
+    with patch(
+        "homeassistant.components.totalconnect.config_flow.TotalConnectConfigFlow.is_valid",
```
As above.

austinmroczek replied:
done

MartinHjelmare commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,105 @@
+
+    # Should fail, same USERNAME (flow)
+    with patch(
+        "homeassistant.components.totalconnect.config_flow.TotalConnectConfigFlow.is_valid",
```
As above.

austinmroczek replied:
done

MartinHjelmare commented on tests/components/totalconnect/test_config_flow.py:
```diff
@@ -0,0 +1,105 @@
+async def test_login_failed(hass):
+    """Test when we have errors during login."""
+    with patch(
+        "homeassistant.components.totalconnect.config_flow.TotalConnectConfigFlow.is_valid",
```
As above.

austinmroczek replied:
done

MartinHjelmare commented:
Please rebase on latest dev branch to let the build pass.

austinmroczek committed:
remove commented code and decorator

austinmroczek committed:
Update tests/components/totalconnect/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

austinmroczek committed:
fix test_config_flow.py

austinmroczek committed:
Merge remote-tracking branch 'upstream/dev' into totalconnect-config-flow

austinmroczek committed:
bump total-connect-client to 0.54

austinmroczek committed:
remove un-needed import of mock_coro

austinmroczek committed:
bump to total-connect-client 0.54.1

austinmroczek committed:
Merge branch 'totalconnect-config-flow' of https://github.com/austinmroczek/home-assistant into totalconnect-config-flow

MartinHjelmare commented on homeassistant/components/totalconnect/__init__.py:
```diff
@@ -1,58 +1,71 @@
+    if DOMAIN not in config:
+        return True
 
-CONFIG_SCHEMA = vol.Schema(
```
Why are we removing the config schema? We're still allowing import of the config to the config flow, so then we still need the config schema.

austinmroczek replied:
Sorry.  I looked at multiple examples given (axis, hue, icloud) and and mine was similar.  I must not understand some inherited properties because CONFIG_SCHEMA is no longer used inside async_setup in any of them. I've re-added it.

MartinHjelmare replied:
`CONFIG_SCHEMA` is used by the core component setup procedure that sets up all components. So it's happening behind the scenes.

austinmroczek committed:
re-add CONFIG_SCHEMA

MartinHjelmare commented:
This is not a breaking change, correct?

austinmroczek commented:
This is not a breaking change.

austinmroczek commented:
I'm getting a pylint error for unused import on config_flow.py.  Even after inserting `# pylint: disable=unused-import` at the end of the line, I still get the error.  Any suggestions?

MartinHjelmare commented:
That's a know pylint bug. Please disable that error locally.

austinmroczek committed:
disable pylint on line 10 to avoid pylint bug

MartinHjelmare reviewed (approved):
Looks good!

austinmroczek commented:
Anything left to do before this is merged?

MartinHjelmare commented:
We should wait with merging this until we know how to best solve the linked issue with regards to pulling the fix into a release.

austinmroczek committed:
Merge branch 'dev' into totalconnect-config-flow

MartinHjelmare commented:
Can be merged when build passes.

austinmroczek committed:
Merge branch 'dev' into totalconnect-config-flow
